File: workers/cpp/worker.cpp
#include <iostream>
#include <string>
#include <thread>
#include <chrono>

// Note: In a real-world scenario, you would use 'hiredis' or 'redis-plus-plus' 
// to connect to Redis and pull jobs. This is a skeleton demonstrating the architecture.
//
// Dependencies to install before compiling:
// brew install hiredis

void processJob(const std::string& taskId, const std::string& payload) {
    std::cout << "[C++ Worker] Processing Heavy Compute Task: " << taskId << std::endl;
    
    // Simulate intense C++ computation
    std::this_thread::sleep_for(std::chrono::seconds(3));
    
    std::cout << "[C++ Worker] Computation for task " << taskId << " finished." << std::endl;

    // TODO: Send HTTP POST to Node.js webhook to mark complete
    // curl -X POST http://localhost:3000/tasks/webhook/<taskId> -H "Content-Type: application/json" -d '{"status":"completed"}'
    std::string curlCmd = "curl -s -X POST http://localhost:3000/tasks/webhook/" + taskId + 
                          " -H \"Content-Type: application/json\" -d '{\"status\":\"completed\"}' > /dev/null";
    system(curlCmd.c_str());
    
    std::cout << "[C++ Worker] Notified Node API." << std::endl;
}

int main() {
    std::cout << "🚀 C++ High-Performance Worker starting..." << std::endl;
    std::cout << "Listening to 'cppQueue' (Simulated)..." << std::endl;

    // Simulated event loop waiting for jobs
    while (true) {
        // In reality, this would be a blocking BRPOP on the Redis list queue
        std::this_thread::sleep_for(std::chrono::seconds(5));
        
        // Simulating receiving a job every 5 seconds for demonstration
        std::string mockTaskId = "task-uuid-cpp-1234";
        std::string mockPayload = "{\"matrix_size\": 1000}";
        
        // processJob(mockTaskId, mockPayload);
        // std::cout << "Waiting for new jobs..." << std::endl;
    }

    return 0;
}

File: workers/python/worker.py
import asyncio
import requests
from bullmq import Worker

# Configuration
REDIS_HOST = 'localhost'
REDIS_PORT = 6379
WEBHOOK_URL = 'http://localhost:3000/tasks/webhook'

async def process_job(job, job_token):
    task_id = job.data.get('id')
    task_type = job.data.get('taskType')
    payload = job.data.get('payload', {})
    
    print(f"[{task_type}] Python Worker processing task {task_id}")
    
    # Simulate heavy processing in Python
    await asyncio.sleep(2)
    
    if task_type == 'PROCESS_DATA':
        print(f"[PROCESS_DATA] Crunching numbers for payload: {payload}")
        # Insert your ML model or data pipeline logic here!
        await asyncio.sleep(3)
        print(f"[PROCESS_DATA] Data crunching complete.")
    else:
        print(f"[{task_type}] Generic processing complete.")
    
    # Notify Node.js API that task is complete
    try:
        response = requests.post(f"{WEBHOOK_URL}/{task_id}", json={"status": "completed"})
        if response.status_code == 200:
            print(f"[Webhook] Successfully notified Node.js API of completion.")
        else:
            print(f"[Webhook] Failed to notify API. Status: {response.status_code}")
    except Exception as e:
        print(f"[Webhook Error] {e}")
        
    return "Done"

async def main():
    print("🚀 Python Worker starting... Listening to 'pythonQueue'")
    
    worker = Worker(
        "pythonQueue",
        process_job,
        {"connection": f"redis://{REDIS_HOST}:{REDIS_PORT}"}
    )
    
    try:
        # Keep worker running
        while True:
            await asyncio.sleep(1)
    except KeyboardInterrupt:
        print("Stopping Python worker...")
        await worker.close()

if __name__ == "__main__":
    asyncio.run(main())
